File: apps/admin-ui/src/api/knowledge.ts
/**
 * SDK for ``/v1/knowledge`` — Stream H.7 + KB commercial uplift.
 *
 * Every endpoint is RAW (no envelope) → ``apiClient`` directly, never
 * ``getJson`` ([memory:envelope-vs-raw-contract-check]).
 *
 * Tenant semantics (Mini-ADR H-19): the backend router reads the JWT's
 * home tenant only — there is no ``tenant_id`` query support, so these
 * calls intentionally take NO TenantScope. Knowledge bases are
 * tenant-scoped shared assets (not per-user).
 */
import { apiClient } from "./client";

/** Document ingestion lifecycle — mirrors the backend ``DocumentStatus``. */
export type DocumentStatus = "pending" | "processing" | "ready" | "failed";

/** Per-base recall strategy — mirrors the backend ``RetrievalMethod``. */
export type RetrievalMethod = "vector" | "keyword" | "hybrid";

/** Per-base retrieval defaults (surfaced so they are not hardcoded). */
export interface RetrievalConfig {
  top_k: number;
  /** Minimum vector similarity to keep a hit; ``null`` disables the cutoff. */
  score_threshold: number | null;
  method: RetrievalMethod;
  rerank_enabled: boolean;
}

export interface KnowledgeBaseStats {
  document_count: number;
  chunk_count: number;
}

export interface KnowledgeBase {
  id: string;
  name: string;
  chunk_max_tokens: number;
  chunk_overlap_tokens: number;
  created_at: string | null;
  // ── commercial-uplift fields (optional so list rows + KnowledgePicker
  //    stubs keep compiling; the backend always returns them). ──
  description?: string | null;
  created_by?: string | null;
  updated_at?: string | null;
  retrieval_config?: RetrievalConfig;
  embedding_provider?: string | null;
  embedding_model?: string | null;
  /** The recorded embedding model differs from the live platform model. */
  needs_reindex?: boolean;
  /** A re-index is in flight (the base is re-embedding its chunk text). */
  reindexing?: boolean;
  stats?: KnowledgeBaseStats;
}

export interface KnowledgeDocument {
  id: string;
  filename: string;
  status: DocumentStatus;
  /** Failure detail when ``status === "failed"``. */
  error: string | null;
  chunk_count: number;
  /** Ingestion attempts so far (durability). */
  attempts?: number;
  created_at: string | null;
  updated_at: string | null;
}

/** One chunk in the segment-preview list (embedding intentionally omitted). */
export interface KnowledgeChunk {
  id: string;
  chunk_index: number;
  content: string;
}

export interface ChunksPage {
  chunks: KnowledgeChunk[];
  total: number;
  offset: number;
  limit: number;
}

/** One ranked hit from the retrieval ("hit testing") endpoint. */
export interface RetrievalTestResult {
  content: string;
  /** ``filename#chunk_index`` source attribution. */
  source: string;
  filename: string;
  chunk_index: number;
  /** Vector cosine similarity in [0, 1]; ``null`` for keyword-only hits. */
  score: number | null;
  /** Which recall path surfaced it: ``vector`` | ``keyword`` | ``both``. */
  recall_source: string | null;
}

export interface RetrievalTestResponse {
  query: string;
  results: RetrievalTestResult[];
  count: number;
}

/** Update payload — only supplied keys are applied; ``null`` clears a
 *  nullable field. ``name`` is intentionally absent (rename is unsupported). */
export interface UpdateBaseBody {
  description?: string | null;
  chunk_max_tokens?: number;
  chunk_overlap_tokens?: number;
  retrieval_top_k?: number;
  retrieval_score_threshold?: number | null;
  retrieval_method?: RetrievalMethod;
  rerank_enabled?: boolean;
}

export interface RetrievalTestBody {
  query: string;
  top_k?: number;
  method?: RetrievalMethod;
  score_threshold?: number | null;
  rerank?: boolean;
}

/** Upload whitelist — mirrors the backend's SUPPORTED_EXTENSIONS
 *  (``knowledge/parsing.py``); used for the Upload ``accept`` attr and
 *  a pre-flight check so unsupported files fail before the request. */
export const SUPPORTED_DOCUMENT_EXTENSIONS = [
  ".pdf",
  ".docx",
  ".pptx",
  ".xlsx",
  ".md",
  ".markdown",
  ".txt",
  ".html",
  ".htm",
  ".csv",
] as const;

export function isSupportedDocument(filename: string): boolean {
  const dot = filename.lastIndexOf(".");
  if (dot < 0) return false;
  const ext = filename.slice(dot).toLowerCase();
  return (SUPPORTED_DOCUMENT_EXTENSIONS as readonly string[]).includes(ext);
}

export async function listBases(): Promise<KnowledgeBase[]> {
  const response = await apiClient.get<{ bases: KnowledgeBase[] }>("/v1/knowledge/bases");
  return response.data.bases;
}

export async function getBase(name: string): Promise<KnowledgeBase> {
  const response = await apiClient.get<KnowledgeBase>(
    `/v1/knowledge/bases/${encodeURIComponent(name)}`,
  );
  return response.data;
}

export async function createBase(params: {
  name: string;
  description?: string;
  chunkMaxTokens?: number;
  chunkOverlapTokens?: number;
  retrievalTopK?: number;
  retrievalScoreThreshold?: number | null;
  retrievalMethod?: RetrievalMethod;
  rerankEnabled?: boolean;
}): Promise<KnowledgeBase> {
  const response = await apiClient.post<KnowledgeBase>("/v1/knowledge/bases", {
    name: params.name,
    description: params.description,
    chunk_max_tokens: params.chunkMaxTokens,
    chunk_overlap_tokens: params.chunkOverlapTokens,
    retrieval_top_k: params.retrievalTopK,
    retrieval_score_threshold: params.retrievalScoreThreshold,
    retrieval_method: params.retrievalMethod,
    rerank_enabled: params.rerankEnabled,
  });
  return response.data;
}

export async function updateBase(name: string, body: UpdateBaseBody): Promise<KnowledgeBase> {
  const response = await apiClient.patch<KnowledgeBase>(
    `/v1/knowledge/bases/${encodeURIComponent(name)}`,
    body,
  );
  return response.data;
}

export async function deleteBase(name: string): Promise<void> {
  await apiClient.delete(`/v1/knowledge/bases/${encodeURIComponent(name)}`);
}

/** Re-embed the base's retained chunk text with the current platform model
 *  (202 Accepted; the base reports ``reindexing`` until it completes). */
export async function reindexBase(name: string): Promise<void> {
  await apiClient.post(`/v1/knowledge/bases/${encodeURIComponent(name)}/reindex`);
}

/** POST multipart — 202 Accepted; ingestion runs in the background and
 *  the caller polls ``listDocuments`` for the status (Mini-ADR H-18). */
export async function uploadDocument(baseName: string, file: File): Promise<KnowledgeDocument> {
  const form = new FormData();
  form.append("file", file);
  const response = await apiClient.post<KnowledgeDocument>(
    `/v1/knowledge/bases/${encodeURIComponent(baseName)}/documents`,
    form,
    { headers: { "Content-Type": "multipart/form-data" } },
  );
  return response.data;
}

export async function listDocuments(baseName: string): Promise<KnowledgeDocument[]> {
  const response = await apiClient.get<{ documents: KnowledgeDocument[] }>(
    `/v1/knowledge/bases/${encodeURIComponent(baseName)}/documents`,
  );
  return response.data.documents;
}

export async function deleteDocument(baseName: string, documentId: string): Promise<void> {
  await apiClient.delete(
    `/v1/knowledge/bases/${encodeURIComponent(baseName)}/documents/${encodeURIComponent(documentId)}`,
  );
}

/** Re-drive a document's ingestion from its retained bytes (202 Accepted).
 *  Throws (HTTP 409) for a legacy document whose bytes were not retained. */
export async function reingestDocument(
  baseName: string,
  documentId: string,
): Promise<KnowledgeDocument> {
  const response = await apiClient.post<KnowledgeDocument>(
    `/v1/knowledge/bases/${encodeURIComponent(baseName)}/documents/${encodeURIComponent(documentId)}/reingest`,
  );
  return response.data;
}

export async function listChunks(
  baseName: string,
  documentId: string,
  params?: { offset?: number; limit?: number },
): Promise<ChunksPage> {
  const response = await apiClient.get<ChunksPage>(
    `/v1/knowledge/bases/${encodeURIComponent(baseName)}/documents/${encodeURIComponent(documentId)}/chunks`,
    { params: { offset: params?.offset, limit: params?.limit } },
  );
  return response.data;
}

/** Retrieval "hit testing" — run a query through the live retrieval pipeline
 *  and see the ranked chunks with scores + recall path. */
export async function testRetrieval(
  baseName: string,
  body: RetrievalTestBody,
): Promise<RetrievalTestResponse> {
  const response = await apiClient.post<RetrievalTestResponse>(
    `/v1/knowledge/bases/${encodeURIComponent(baseName)}/test`,
    body,
  );
  return response.data;
}

File: services/control-plane/src/control_plane/api/knowledge.py
"""``/v1/knowledge`` — Stream J.5 knowledge-base management + document ingest.

Knowledge bases are tenant-scoped (shared, not per-user). This router
manages bases and their documents:

* bases — create / list / delete;
* documents — upload (async ingest) / list (with status) / delete.

Upload is **off the request path**: the document is recorded ``pending``
and handed to the :class:`KnowledgeIngestionRunner`; the caller polls the
document list for its ``status``.
"""

from __future__ import annotations

import hashlib
import logging
from datetime import datetime
from pathlib import Path
from typing import Annotated, Any
from uuid import UUID

from fastapi import APIRouter, Depends, File, HTTPException, Request, Response, UploadFile
from fastapi.responses import JSONResponse
from pydantic import BaseModel, Field

from control_plane.knowledge.ingestion import KnowledgeIngestionRunner
from control_plane.knowledge.parsing import SUPPORTED_EXTENSIONS
from helix_agent.persistence import KnowledgeStore
from helix_agent.persistence.knowledge import UNSET, DuplicateKnowledgeBaseError
from helix_agent.protocol import (
    DEFAULT_CHUNK_MAX_TOKENS,
    DEFAULT_CHUNK_OVERLAP_TOKENS,
    DEFAULT_RETRIEVAL_TOP_K,
    KnowledgeBase,
    KnowledgeDocument,
    RetrievalMethod,
)

# NOTE: ``KnowledgeRetriever`` lives in ``orchestrator`` — importing it at
# module top level would cycle (control-plane → orchestrator). The
# retrieval-test endpoint depends on the object duck-typed (``Any``); it is
# the same instance the lifespan builds and stashes on ``app.state``.

logger = logging.getLogger("helix.control_plane.knowledge")


class _CreateBaseBody(BaseModel):
    """Body of ``POST /v1/knowledge/bases``."""

    name: str = Field(min_length=1, max_length=128)
    description: str | None = Field(default=None, max_length=2000)
    chunk_max_tokens: int | None = Field(default=None, gt=0)
    chunk_overlap_tokens: int | None = Field(default=None, ge=0)
    retrieval_top_k: int | None = Field(default=None, ge=1, le=50)
    retrieval_score_threshold: float | None = Field(default=None, ge=0.0, le=1.0)
    retrieval_method: RetrievalMethod | None = None
    rerank_enabled: bool | None = None


class _UpdateBaseBody(BaseModel):
    """Body of ``PATCH /v1/knowledge/bases/{name}``.

    Only the fields the caller actually sends are applied (decided via
    ``model_fields_set``), so a nullable field can be cleared with an
    explicit ``null`` and left alone by omission. ``name`` is intentionally
    absent — renaming would silently break agent ``knowledge_base_refs``.
    """

    description: str | None = Field(default=None, max_length=2000)
    chunk_max_tokens: int | None = Field(default=None, gt=0)
    chunk_overlap_tokens: int | None = Field(default=None, ge=0)
    retrieval_top_k: int | None = Field(default=None, ge=1, le=50)
    retrieval_score_threshold: float | None = Field(default=None, ge=0.0, le=1.0)
    retrieval_method: RetrievalMethod | None = None
    rerank_enabled: bool | None = None


class _TestBody(BaseModel):
    """Body of ``POST /v1/knowledge/bases/{name}/test`` (retrieval hit-test)."""

    query: str = Field(min_length=1, max_length=4000)
    top_k: int | None = Field(default=None, ge=1, le=50)
    method: RetrievalMethod | None = None
    score_threshold: float | None = Field(default=None, ge=0.0, le=1.0)
    rerank: bool | None = None


def _get_knowledge_store(request: Request) -> KnowledgeStore:
    return request.app.state.knowledge_store  # type: ignore[no-any-return]


def _get_ingestion_runner(request: Request) -> KnowledgeIngestionRunner | None:
    return request.app.state.knowledge_ingestion_runner  # type: ignore[no-any-return]


def _get_knowledge_retriever(request: Request) -> Any | None:
    return getattr(request.app.state, "knowledge_retriever", None)


def _get_embedding_config_service(request: Request) -> Any | None:
    return getattr(request.app.state, "platform_embedding_config_service", None)


async def _current_embedding_model(config_service: Any | None) -> tuple[str, str] | None:
    if config_service is None:
        return None
    result: tuple[str, str] | None = await config_service.effective_embedding_config()
    return result


def _needs_reindex(base: KnowledgeBase, current: tuple[str, str] | None) -> bool:
    """A base needs re-indexing when its recorded embedding model differs from
    the live platform model. Unknown on either side (legacy base / unconfigured
    platform) → ``False`` (nothing to compare)."""
    if base.embedding_model is None or current is None:
        return False
    return (base.embedding_provider, base.embedding_model) != current


def _iso(value: datetime | None) -> str | None:
    return value.isoformat() if value is not None else None


def _base_dict(
    base: KnowledgeBase,
    stats: tuple[int, int] = (0, 0),
    *,
    needs_reindex: bool = False,
) -> dict[str, Any]:
    document_count, chunk_count = stats
    return {
        "id": str(base.id),
        "name": base.name,
        "description": base.description,
        "created_by": base.created_by,
        "chunk_max_tokens": base.chunk_max_tokens,
        "chunk_overlap_tokens": base.chunk_overlap_tokens,
        "retrieval_config": {
            "top_k": base.retrieval_top_k,
            "score_threshold": base.retrieval_score_threshold,
            "method": base.retrieval_method.value,
            "rerank_enabled": base.rerank_enabled,
        },
        "embedding_provider": base.embedding_provider,
        "embedding_model": base.embedding_model,
        "needs_reindex": needs_reindex,
        "reindexing": base.reindex_requested_at is not None,
        "stats": {"document_count": document_count, "chunk_count": chunk_count},
        "created_at": _iso(base.created_at),
        "updated_at": _iso(base.updated_at),
    }


def _document_dict(document: KnowledgeDocument) -> dict[str, Any]:
    return {
        "id": str(document.id),
        "filename": document.filename,
        "status": document.status.value,
        "error": document.error,
        "chunk_count": document.chunk_count,
        "attempts": document.attempts,
        "created_at": _iso(document.created_at),
        "updated_at": _iso(document.updated_at),
    }


async def _require_base(store: KnowledgeStore, tenant_id: UUID, name: str) -> KnowledgeBase:
    base = await store.get_base(tenant_id=tenant_id, name=name)
    if base is None:
        raise HTTPException(status_code=404, detail="knowledge base not found")
    return base


def build_knowledge_router() -> APIRouter:
    router = APIRouter(prefix="/v1/knowledge", tags=["knowledge"])

    @router.post("/bases", response_model=None)
    async def create_base(
        body: _CreateBaseBody,
        request: Request,
        store: Annotated[KnowledgeStore, Depends(_get_knowledge_store)],
        config_service: Annotated[Any | None, Depends(_get_embedding_config_service)],
    ) -> JSONResponse:
        tenant_id: UUID = request.state.tenant_id
        max_tokens = body.chunk_max_tokens or DEFAULT_CHUNK_MAX_TOKENS
        overlap_tokens = (
            body.chunk_overlap_tokens
            if body.chunk_overlap_tokens is not None
            else DEFAULT_CHUNK_OVERLAP_TOKENS
        )
        if overlap_tokens >= max_tokens:
            raise HTTPException(
                status_code=400,
                detail="chunk_overlap_tokens must be less than chunk_max_tokens",
            )
        # Pin the live platform embedding model so a later model swap is
        # detectable (``needs_reindex``). NULL when embedding is unconfigured.
        current = await _current_embedding_model(config_service)
        provider, model = current if current is not None else (None, None)
        try:
            base = await store.create_base(
                tenant_id=tenant_id,
                name=body.name,
                description=body.description,
                created_by=getattr(request.state, "actor_id", None),
                chunk_max_tokens=max_tokens,
                chunk_overlap_tokens=overlap_tokens,
                retrieval_top_k=body.retrieval_top_k or DEFAULT_RETRIEVAL_TOP_K,
                retrieval_score_threshold=body.retrieval_score_threshold,
                retrieval_method=body.retrieval_method or RetrievalMethod.HYBRID,
                rerank_enabled=body.rerank_enabled if body.rerank_enabled is not None else True,
                embedding_provider=provider,
                embedding_model=model,
            )
        except DuplicateKnowledgeBaseError as exc:
            raise HTTPException(status_code=409, detail="knowledge base already exists") from exc
        return JSONResponse(
            status_code=201, content=_base_dict(base, needs_reindex=_needs_reindex(base, current))
        )

    @router.get("/bases", response_model=None)
    async def list_bases(
        request: Request,
        store: Annotated[KnowledgeStore, Depends(_get_knowledge_store)],
        config_service: Annotated[Any | None, Depends(_get_embedding_config_service)],
    ) -> JSONResponse:
        tenant_id: UUID = request.state.tenant_id
        bases = await store.list_bases(tenant_id=tenant_id)
        stats = await store.base_stats_many(tenant_id=tenant_id)
        current = await _current_embedding_model(config_service)
        return JSONResponse(
            content={
                "bases": [
                    _base_dict(
                        base,
                        stats.get(base.id, (0, 0)),
                        needs_reindex=_needs_reindex(base, current),
                    )
                    for base in bases
                ]
            }
        )

    @router.get("/bases/{name}", response_model=None)
    async def get_base(
        name: str,
        request: Request,
        store: Annotated[KnowledgeStore, Depends(_get_knowledge_store)],
        config_service: Annotated[Any | None, Depends(_get_embedding_config_service)],
    ) -> JSONResponse:
        tenant_id: UUID = request.state.tenant_id
        base = await _require_base(store, tenant_id, name)
        stats = await store.base_stats(tenant_id=tenant_id, kb_id=base.id)
        current = await _current_embedding_model(config_service)
        return JSONResponse(
            content=_base_dict(base, stats, needs_reindex=_needs_reindex(base, current))
        )

    @router.patch("/bases/{name}", response_model=None)
    async def update_base(
        name: str,
        body: _UpdateBaseBody,
        request: Request,
        store: Annotated[KnowledgeStore, Depends(_get_knowledge_store)],
        config_service: Annotated[Any | None, Depends(_get_embedding_config_service)],
    ) -> JSONResponse:
        tenant_id: UUID = request.state.tenant_id
        base = await _require_base(store, tenant_id, name)
        sent = body.model_fields_set
        # Resolve the effective overlap/max for the cross-field check, falling
        # back to the stored values for whichever side the caller omitted (or
        # sent as null — these columns are non-nullable, so null = leave alone).
        new_max = (
            body.chunk_max_tokens
            if "chunk_max_tokens" in sent and body.chunk_max_tokens is not None
            else base.chunk_max_tokens
        )
        new_overlap = (
            body.chunk_overlap_tokens
            if "chunk_overlap_tokens" in sent and body.chunk_overlap_tokens is not None
            else base.chunk_overlap_tokens
        )
        if new_overlap >= new_max:
            raise HTTPException(
                status_code=400,
                detail="chunk_overlap_tokens must be less than chunk_max_tokens",
            )
        updated = await store.update_base(
            tenant_id=tenant_id,
            kb_id=base.id,
            description=body.description if "description" in sent else UNSET,
            chunk_max_tokens=body.chunk_max_tokens if "chunk_max_tokens" in sent else None,
            chunk_overlap_tokens=(
                body.chunk_overlap_tokens if "chunk_overlap_tokens" in sent else None
            ),
            retrieval_top_k=body.retrieval_top_k if "retrieval_top_k" in sent else None,
            retrieval_score_threshold=(
                body.retrieval_score_threshold if "retrieval_score_threshold" in sent else UNSET
            ),
            retrieval_method=body.retrieval_method if "retrieval_method" in sent else None,
            rerank_enabled=body.rerank_enabled if "rerank_enabled" in sent else None,
        )
        if updated is None:  # pragma: no cover - guarded by _require_base above
            raise HTTPException(status_code=404, detail="knowledge base not found")
        stats = await store.base_stats(tenant_id=tenant_id, kb_id=updated.id)
        current = await _current_embedding_model(config_service)
        return JSONResponse(
            content=_base_dict(updated, stats, needs_reindex=_needs_reindex(updated, current))
        )

    @router.post("/bases/{name}/reindex", response_model=None)
    async def reindex_base(
        name: str,
        request: Request,
        store: Annotated[KnowledgeStore, Depends(_get_knowledge_store)],
        runner: Annotated[KnowledgeIngestionRunner | None, Depends(_get_ingestion_runner)],
        config_service: Annotated[Any | None, Depends(_get_embedding_config_service)],
    ) -> JSONResponse:
        tenant_id: UUID = request.state.tenant_id
        base = await _require_base(store, tenant_id, name)
        if runner is None:
            raise HTTPException(
                status_code=503,
                detail="knowledge ingestion unavailable: no embedding model configured",
            )
        current = await _current_embedding_model(config_service)
        if current is None:
            raise HTTPException(
                status_code=503,
                detail="knowledge ingestion unavailable: no embedding model configured",
            )
        provider, model = current
        # Mark in-flight (UI shows "re-indexing"); the runner re-embeds retained
        # chunk text, stamps the model, and clears the flag on completion.
        await store.request_reindex(tenant_id=tenant_id, kb_id=base.id)
        runner.submit_reindex(
            tenant_id=tenant_id,
            kb_id=base.id,
            embedding_provider=provider,
            embedding_model=model,
        )
        return JSONResponse(status_code=202, content={"status": "reindexing", "name": base.name})

    @router.delete("/bases/{name}", status_code=204, response_model=None)
    async def delete_base(
        name: str,
        request: Request,
        store: Annotated[KnowledgeStore, Depends(_get_knowledge_store)],
    ) -> Response:
        tenant_id: UUID = request.state.tenant_id
        base = await _require_base(store, tenant_id, name)
        await store.delete_base(tenant_id=tenant_id, kb_id=base.id)
        return Response(status_code=204)

    @router.post("/bases/{name}/documents", response_model=None)
    async def upload_document(
        name: str,
        request: Request,
        file: Annotated[UploadFile, File()],
        store: Annotated[KnowledgeStore, Depends(_get_knowledge_store)],
        runner: Annotated[KnowledgeIngestionRunner | None, Depends(_get_ingestion_runner)],
    ) -> JSONResponse:
        tenant_id: UUID = request.state.tenant_id
        base = await _require_base(store, tenant_id, name)
        if runner is None:
            raise HTTPException(
                status_code=503,
                detail="knowledge ingestion unavailable: no embedding model configured",
            )
        filename = file.filename
        if not filename:
            raise HTTPException(status_code=400, detail="uploaded file has no filename")
        if Path(filename).suffix.lower() not in SUPPORTED_EXTENSIONS:
            raise HTTPException(status_code=400, detail=f"unsupported document type: {filename!r}")
        raw = await file.read()
        # Retain the original bytes so a crashed/failed document can be
        # re-driven (crash recovery + re-ingest) without a re-upload.
        document = await store.upsert_document(
            tenant_id=tenant_id,
            kb_id=base.id,
            filename=filename,
            content=raw,
            content_sha256=hashlib.sha256(raw).hexdigest(),
        )
        runner.submit(
            tenant_id=tenant_id,
            document_id=document.id,
            kb_id=base.id,
            filename=filename,
            raw=raw,
            chunk_max_tokens=base.chunk_max_tokens,
            chunk_overlap_tokens=base.chunk_overlap_tokens,
        )
        # 202 Accepted — ingestion runs in the background; poll the
        # document list for its status.
        return JSONResponse(status_code=202, content=_document_dict(document))

    @router.get("/bases/{name}/documents", response_model=None)
    async def list_documents(
        name: str,
        request: Request,
        store: Annotated[KnowledgeStore, Depends(_get_knowledge_store)],
    ) -> JSONResponse:
        tenant_id: UUID = request.state.tenant_id
        base = await _require_base(store, tenant_id, name)
        documents = await store.list_documents(tenant_id=tenant_id, kb_id=base.id)
        return JSONResponse(content={"documents": [_document_dict(doc) for doc in documents]})

    @router.delete("/bases/{name}/documents/{document_id}", status_code=204, response_model=None)
    async def delete_document(
        name: str,
        document_id: UUID,
        request: Request,
        store: Annotated[KnowledgeStore, Depends(_get_knowledge_store)],
    ) -> Response:
        tenant_id: UUID = request.state.tenant_id
        await _require_base(store, tenant_id, name)
        deleted = await store.delete_document(tenant_id=tenant_id, document_id=document_id)
        if not deleted:
            raise HTTPException(status_code=404, detail="document not found")
        return Response(status_code=204)

    @router.post("/bases/{name}/documents/{document_id}/reingest", response_model=None)
    async def reingest_document(
        name: str,
        document_id: UUID,
        request: Request,
        store: Annotated[KnowledgeStore, Depends(_get_knowledge_store)],
        runner: Annotated[KnowledgeIngestionRunner | None, Depends(_get_ingestion_runner)],
    ) -> JSONResponse:
        tenant_id: UUID = request.state.tenant_id
        base = await _require_base(store, tenant_id, name)
        if runner is None:
            raise HTTPException(
                status_code=503,
                detail="knowledge ingestion unavailable: no embedding model configured",
            )
        document = await store.get_document(tenant_id=tenant_id, document_id=document_id)
        if document is None or document.kb_id != base.id:
            raise HTTPException(status_code=404, detail="document not found")
        content = await store.get_document_content(tenant_id=tenant_id, document_id=document_id)
        if content is None:
            # Legacy document predating retained bytes — cannot re-drive.
            raise HTTPException(
                status_code=409,
                detail="original file not retained; please re-upload this document",
            )
        # Reset to pending (keeps the same bytes) and re-drive the fast path.
        reset = await store.upsert_document(
            tenant_id=tenant_id,
            kb_id=base.id,
            filename=document.filename,
            content=content,
            content_sha256=hashlib.sha256(content).hexdigest(),
        )
        runner.submit(
            tenant_id=tenant_id,
            document_id=reset.id,
            kb_id=base.id,
            filename=document.filename,
            raw=content,
            chunk_max_tokens=base.chunk_max_tokens,
            chunk_overlap_tokens=base.chunk_overlap_tokens,
        )
        return JSONResponse(status_code=202, content=_document_dict(reset))

    @router.get("/bases/{name}/documents/{document_id}/chunks", response_model=None)
    async def list_chunks(
        name: str,
        document_id: UUID,
        request: Request,
        store: Annotated[KnowledgeStore, Depends(_get_knowledge_store)],
        offset: int = 0,
        limit: int = 50,
    ) -> JSONResponse:
        tenant_id: UUID = request.state.tenant_id
        base = await _require_base(store, tenant_id, name)
        document = await store.get_document(tenant_id=tenant_id, document_id=document_id)
        if document is None or document.kb_id != base.id:
            raise HTTPException(status_code=404, detail="document not found")
        safe_offset = max(0, offset)
        safe_limit = max(1, min(limit, 200))
        chunks, total = await store.list_chunks(
            tenant_id=tenant_id,
            document_id=document_id,
            offset=safe_offset,
            limit=safe_limit,
        )
        return JSONResponse(
            content={
                "chunks": [
                    {"id": str(c.id), "chunk_index": c.chunk_index, "content": c.content}
                    for c in chunks
                ],
                "total": total,
                "offset": safe_offset,
                "limit": safe_limit,
            }
        )

    @router.post("/bases/{name}/test", response_model=None)
    async def test_retrieval(
        name: str,
        body: _TestBody,
        request: Request,
        store: Annotated[KnowledgeStore, Depends(_get_knowledge_store)],
        retriever: Annotated[Any | None, Depends(_get_knowledge_retriever)],
    ) -> JSONResponse:
        tenant_id: UUID = request.state.tenant_id
        base = await _require_base(store, tenant_id, name)
        if retriever is None:
            raise HTTPException(
                status_code=503,
                detail="knowledge retrieval unavailable: no embedding model configured",
            )
        results = await retriever.search(
            tenant_id=tenant_id,
            base_names=[base.name],
            query=body.query,
            limit=body.top_k or base.retrieval_top_k,
            method=body.method,
            score_threshold=body.score_threshold,
            rerank=body.rerank,
        )
        return JSONResponse(
            content={
                "query": body.query,
                "results": [
                    {
                        "content": r.content,
                        "source": f"{r.filename}#{r.chunk_index}",
                        "filename": r.filename,
                        "chunk_index": r.chunk_index,
                        "score": r.score,
                        "recall_source": r.recall_source,
                    }
                    for r in results
                ],
                "count": len(results),
            }
        )

    return router
